### PR TITLE
AC_FENCE:Add AUTOENABLE deprecation warning

### PR DESCRIPTION
**Pull request by Hwurzburg**
Can be changed to prearm checkfailed in 4.7

### CODE REVIEW

File: libraries/AC_Fence/AC_Fence.cpp [hunk 1]
```diff
@@ -429,6 +429,16 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
         hal.util->snprintf(failure_msg, failure_msg_len, "Fences enabled, but none selected");
         return false;
     }
+    
+    // if AUTOENABLE = 1 or 2 warn now, but fail in a later release
+    static uint8_t i = 0 ;
+    i++;
+    if (i > 60) {
```
Comment (peterbarker) [inline libraries/AC_Fence/AC_Fence.cpp:436]:
```suggestion
    // PARAMETER_CONVERSION - Added: Jul-2024 for ArduPilot-4.6
    const uint32_t now_ms = AP_HAL::millis();
    if (now_ms - last_autoenable_warn_ms > 60000) {
```

... and add last_autoenable_warn_ms in the header

File: libraries/AC_Fence/AC_Fence.cpp [hunk 2]
```diff
@@ -429,6 +429,16 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
...
+        if (_auto_enabled == 1 or _auto_enabled == 2) {
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FENCE_AUTOENABLE set to 1 or 2, will be deprecated next release, use 0 or 3");
+            i = 0;
```
Comment (peterbarker) [inline libraries/AC_Fence/AC_Fence.cpp:439]:
```suggestion
```

can't use a counter to decimate as this can be called at variable rates.

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AC_Fence/AC_Fence.cpp [hunk 3]
```diff
@@ -429,6 +429,17 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
...
         return false;
     }
+    
+    // if AUTOENABLE = 1 or 2 warn now, but fail in a later release
+    // PARAMETER_CONVERSION - Added: Jul-2024 for ArduPilot-4.6
+    const uint32_t now_ms = AP_HAL::millis();
+    static uint32_t _last_autoenable_warn_ms = now_ms - 60000; //needs this static since function is const
+    if (now_ms - _last_autoenable_warn_ms > 60000) {
+        if (_auto_enabled == 1 or _auto_enabled == 2) {
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FENCE_AUTOENABLE set to 1 or 2, will be deprecated next release, use 0 or 3");
```
Comment (peterbarker) [inline libraries/AC_Fence/AC_Fence.cpp:439]:
```suggestion
            GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FENCE_AUTOENABLE is %u, will be deprecated next release, use 0 or 3", unsigned(_auto_enabled));
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
Commit messageshould be `AC_Fence`., not AC_FENCE

File: libraries/AC_Fence/AC_Fence.cpp [hunk 4]
```diff
@@ -429,6 +429,17 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
...
         return false;
     }
+    
+    // if AUTOENABLE = 1 or 2 warn now, but fail in a later release
+    // PARAMETER_CONVERSION - Added: Jul-2024 for ArduPilot-4.6
+    const uint32_t now_ms = AP_HAL::millis();
+    static uint32_t _last_autoenable_warn_ms = now_ms - 60000; //needs this static since function is const
+    if (now_ms - _last_autoenable_warn_ms > 60000) {
+        if (_auto_enabled == 1 or _auto_enabled == 2) {
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FENCE_AUTOENABLE is %u, will be deprecated in 4.7, use 3", unsigned(_auto_enabled));
```
Comment (peterbarker) [inline libraries/AC_Fence/AC_Fence.cpp:439]:
```suggestion
            GCS_SEND_TEXT(MAV_SEVERITY_WARNING, "FENCE_AUTOENABLE is %u, will be removed in 4.7, use 3", unsigned(_auto_enabled));
```

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AC_Fence/AC_Fence.cpp [hunk 5]
```diff
@@ -429,6 +429,17 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
         hal.util->snprintf(failure_msg, failure_msg_len, "Fences enabled, but none selected");
         return false;
     }
+    
+    // if AUTOENABLE = 1 or 2 warn now, but fail in a later release
+    // PARAMETER_CONVERSION - Added: Jul-2024 for ArduPilot-4.6
+    const uint32_t now_ms = AP_HAL::millis();
+    static uint32_t _last_autoenable_warn_ms = now_ms - 60000; //needs this static since function is const
```
Comment (andyp1per) [inline libraries/AC_Fence/AC_Fence.cpp:436]: These timings seem fairly arbitrary and pre-arm already has its own timings for output. I would just make this a one-line message and print every time.

**Review by andyp1per — CHANGES_REQUESTED** [review]
I think this could be simpler, but I would also at least like to have the discussion about these options first before simply deprecating them. They were put in for a reason so let's at least make sure that that reason is no longer valid.

File: libraries/AC_Fence/AC_Fence.cpp (hunk 5 above)
Comment (Hwurzburg) [reply libraries/AC_Fence/AC_Fence.cpp:436]: its not a pre-arm yet and we dont want to spam at 30 second rate, nor to make it a one off thats missed

**Comment by Hwurzburg** [comment]
Tridge asked for this....the autoenable was a series of patches to correct issues with fence as far as I can tell....
1 was to allow takeoffs when fences would otherwise interfere if enabled

but then 2 was needed when alt min was added to allow landings...

3 was the only really correct way....

File: libraries/AC_Fence/AC_Fence.cpp [hunk 6]
```diff
@@ -429,6 +429,17 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
...
+    if (now_ms - _last_autoenable_warn_ms > 60000) {
+        if (_auto_enabled == 1 or _auto_enabled == 2) {
```
Comment (tridge) [inline libraries/AC_Fence/AC_Fence.cpp]: should be || not or

Comment (tridge) [inline libraries/AC_Fence/AC_Fence.cpp]: if for the setting should be on outside

File: libraries/AC_Fence/AC_Fence.cpp (hunk 5 above)
Comment (tridge) [inline libraries/AC_Fence/AC_Fence.cpp]: the static var should not be initialised at all

**Review by tridge — APPROVED** [review]

**Review by andyp1per — APPROVED** [review]

**Comment by Hwurzburg** [comment]
@peterbarker merge on CI pass?

**Pull request merged**